### ISSUE TITLE
Accuracy in TensorFlow Lab Question 3 capped at 0.10266666859388351

### ISSUE THREAD

**Issue by mgruben** [udacity/sdc-issue-reports#110]
This is basically issue #57.

I'm optimizing the three configurations in the Notebook, but never getting above `0.10266666859388351` for accuracy
    The full results are as follows:

```
Configuration 1
    2000    0.09826666861772537
    1000    0.09600000083446503
    500     0.09826666861772537
    300     0.10000000149011612
    50      0.09719999879598618
Configuration 2
    0.8     0.10266666859388351
    0.5     0.10266666859388351
    0.1     0.10266666859388351
    0.05    0.09719999879598618
    0.01    0.09719999879598618
Configuration 3
    1       0.10266666859388351
    2       0.10266666859388351
    3       0.10266666859388351
    4       0.10266666859388351
    5       0.10266666859388351
```

Similar results persist even after restarting the kernel and clearing the output

My repo is [linked](https://github.com/mgruben/TensorLab).

**Comment by mgruben** [comment]
Forgot to add, I'm running Arch Linux 4.8.4 on an HP Envy 17T-s000, and accessing the notebook through chromium 54.0.2840.71-1

**Comment by Brok-Bucholtz** [comment]
@mgruben 

Can you commit your "notMNIST.pickle" file to the repo for me?

**Comment by WuStangDan** [comment]
I'm getting very similar results to @mgruben. Specifically there is almost no change to the validation accuracy regardless of what I use for each parameter.

**Comment by WuStangDan** [comment]
@Brok-Bucholtz So I deleted my pickle file and the downloaded notMNIST files, and ran through all the sections again and now my validation accuracy is moving.

I'm not sure what happened to my first pickle file (maybe I re-ran one of the steps multiple times?) but that was definitely the source of the problem.

**Comment by Brok-Bucholtz** [comment]
@WuStangDan @mgruben

The lab has 10 classes, so getting an accuracy of 0.1 is just the neural network randomly guessing which is the correct label.  This is either due to a bad model or bad training data.  Since the lab provides most of the model, it's probably bad training data.  Deleting the pickle file and rebuilding it should fix most problems with bad training data.

Before you delete the file, please send me a link to your pickle file so I can analyze it and prevent other student from making the same mistake.

**Comment by WuStangDan** [comment]
@Brok-Bucholtz Here is a link to that bad pickle file.

https://drive.google.com/file/d/0B2CMow4Eoy57NUkwcl9VWjB1S1U/view?usp=sharing

**Comment by Brok-Bucholtz** [comment]
@WuStangDan 

Thank You!

**Comment by mgruben** [comment]
@Brok-Bucholtz I've also shared my `notMNIST.pickle `.

https://drive.google.com/open?id=0B-_GN-FxgDU1RVBnenJaZ29uTDg

**Comment by mgruben** [comment]
like @WuStangDan, I:
(1) Deleted my previous `notMNIST.pickle`
(2) ran `jupyter notebook` in the `CarND-TensorFlow-Lab` directory
(3) clicked on `lab.ipynb` to start the lab
(4) restarted the kernel and cleared all output
(5) ran all code cells in order (including re-loading the `pickle` file)
(6) used `epochs = 1, batch_size = 50, learning_rate = 0.1` for the loss-accuracy code cell in Problem 3

This yielded a much more sensible `Validation accuracy at 0.7613333463668823`.

I consider deleting the `pickle` file and re-running all code cells (maybe could skip reading the `pickle` though?) to be a fix for this issue.

**Comment by giladgressel** [comment]
Hey @Brok-Bucholtz 
I had a student with the same problem today, I will ask him to post here

**Comment by rush00121** [comment]
I had the same problem . I had to delete the noMNSIT data files from the directory and rebuilt the whole project. 
Make sure that you do not rebuild each section more than once. 
Then it worked. I guess there was some stale file sitting which caused the model to misbehave.

**Comment by Brok-Bucholtz** [comment]
@mgruben @rush00121 @WuStangDan

Thank you for all your help!  I'll update this issue with the fix, I'm hoping to get it out tomorrow.

**Comment by Brok-Bucholtz** [comment]
@mgruben @rush00121 @WuStangDan 

I added this to our sprint backlog. I'll update this issue when it's fixed.

**Comment by Brok-Bucholtz** [comment]
@mgruben @rush00121 @WuStangDan

I found out the problem.  Both of you ran the normalization twice on the features.  I added a check, to prevent this from happening in the future.  Again, thanks for submitting your pickle files.  That was really helpful!

**Comment by giladgressel** [comment]
Hey Brok, 

Shooot, Bharat and I saw this awhile ago and forgot to tell you.. man I'm sorry.

In general the notebook shouldn't re-use variable names when making transformations.  If students rerun the code this will mess things up, as you've noticed.

Your fix works
```
if not is_features_normal:
    train_features = normalize_greyscale(train_features)
    test_features = normalize_greyscale(test_features)
    is_features_normal = True
```

However you may want to consider adopting an approach that is something like this:

` train_features = normalize_greyscale(non_normalized_features)`

Of course you'd need to initialize `non_normalized_features` in the first place.

Another place in the lab that has this problem is here:
```
# Get randomized datasets for training and validation
train_features, valid_features, train_labels, valid_labels = train_test_split(
    train_features,
    train_labels,
    test_size=0.05,
    random_state=832289)
```
If a student re-runs this cell, they are going to keep shrinking their dataset over and over again.  You should rename either the output variable to be `train_features_split` or something that indicates it's not the original training features.  This way it will initialize a new variable. 

Jupyter is tricky because you can re-run code all the time, so you should never use the same variable name as output if re-running that line of code could cause problems

**Comment by mgruben** [comment]
@Brok-Bucholtz 
Doh!
Thanks for figuring this out :)

**Comment by kerrickstaley** [comment]
This is still happening.

When I run cells from Problem 3 in the lab, sometimes (maybe 30% of the time) the "Loss" graph will be empty (no data), the "Accuracy" graph will be a flat line, and the final "Validation Accuracy" will be some low value close to zero:

![screenshot from 2017-10-15 22-24-53](https://user-images.githubusercontent.com/184773/31596887-261149ee-b1fa-11e7-9bc5-86a209f3808a.png)

The final validation accuracy is always the same value when this happens.

Deleting the notMnist.pickle file and re-running all the cells doesn't fix the problem (although the validation accuracy changes to a new value, also close to 10%, that again doesn't change across runs).

The work-around is to just re-run the cell.

Let me know if you'd like my code or more info about my system configuration.
